README.md
fhem-tablet-ui
========

UI builder framework for FHEM — http://fhem.de/fhem.html
with a clear intention: Keep it short and simple!


![](http://knowthelist.github.io/fhem-tablet-ui/fhem-tablet-ui-example_new.png)

Requires
-------
* jQuery v1.7+
* font-awesome http://fortawesome.github.io/Font-Awesome
* jquery.gridster  http://gridster.net
* jquery.toast
* and more
All requires are included in the lib folder.

Install
-------
 * copy the whole tree into the corresponding folder of your FHEM server /\<fhem-path\>/www/tablet
 * rename the index-example.html to index.html or create your own index.html
 * Tadaaa! A new fhem ui in http://\<fhem-url\>:8083/fhem/tablet/index.html
 
 or just use 'update all https://raw.githubusercontent.com/knowthelist/fhem-tablet-ui/master/controls_fhemtabletui.txt'
 on the FHEM commandline (or input field of FHEMWEB)

Demos
-------
[Widgets](http://knowthelist.github.io/fhem/tablet/demo_widgets.html)

Forum
-------
[FHEM Forum FTUI](https://forum.fhem.de/index.php/board,71.0.html)

Configure
-------
Just configure the **index.html** to change the dashboard for your needs.

Change the wiget container according your rooms
```html
<li data-row="2" data-col="2" data-sizey="3" data-sizex="3">
    <header>KUECHE</header>
    <div class="sheet">
        <div class="row">
            <div class="cell" data-type="thermostat" data-device="KuecheHeizung_Clima"></div>
            <div class="cell">
                <div data-type="switch" data-device="HerdLicht_Sw"></div>
                <div>HerdLicht</div>
            </div>
        </div>
        <div class="row">
            <div class="cell" data-type="thermostat" data-device="KuecheHeizung2_Clima"></div>
            <div class="cell" data-type="symbol" data-device="KuechenFenster"></div>
        </div>
    </div>
</li>
```
Change the widgets you have and want to see on the dashboard
```html
<div data-type="thermostat" data-device='WohnzimmerHeizung_Clima'></div>
```

Widgets
-------
Currently there are more then 20 types of widgets in the base installation.
- **thermostat** : dial for heater thermostates to set desired value and show current value
- **switch** : Toggle any command to FHEM (e.g. on / off)
- **label** : show state as text (colourable)
- **symbol** : show state as an icon (e.g. window open) 
- **push** : send one single command to FHEM e.g. up / down
- **volume** : dial to set a single value (e.g. 0-60)
- **homestatus** : selector for 4 states (1=home,2=night,3=away,4=holiday) 
- **dimmer** : toogle button with a setter for on value
- **slider** : vertical/horizontal slider to select between min/max value
- **image** : insert an image, the URL is given by a reading
- **weather** : insert an icon or image, represending a weather literal
- **circlemenu** : Cover multiple widgets behind a single widget
- **select**	: Combobox to provide a list for selection
- **pagetab**	 : Element to smoothly exchange the whole page with another page
- **pagebutton** : Simple element to jump to another page and is shown as ON on destination Page
- **level** : vertical/horizontal bar to show values between min/max value
- **rotor** : slider to switch automatically between multiple widgets at one position
- **swiper** : touch slider for multiple widgets at one position
- **progress** : round symbolic display  for percent values
- **simplechart** : simple XY line chart for one value (reads directly from fhem log file)
- **popup** : a popup dialog which open on click on another widget 
- **readingsgroup** : displaying a readingsgroup defined in fhem
- **datetimepicker** : select a date and time value from calendar
- **eventmonitor**: for debugging - shows all events which normal widgets 'see'
- **chart** : multistyle chart for multiple values (reads directly from FHEM log file)
- **highchart** : multistyle chart for multiple values
- **checkbox** : Toggle any command to FHEM (e.g. on / off)
- **range** : vertical bar graph to show a values range between min/max value and high/low limits with different colors
- **colorwheel** : select a color from a wheel
- **link** : link or button link element
- **spinner** : is a control element to adjust a value by clicking on the up or down icon
- **departure** : a pulic transport departure schedule widget
- **slideout** : a slide out menu plugin, usable for mobile phone pages.
- **medialist** : a list of media elements e.g. songs in a playlist
- **notify** : popup browser notifications for reading changes
- **theme**  : plugin to actives css stylesheet themes


By default the ui gets/sets the fhem parameter 'STATE' (not 'state').

All widgets have individual parameter settings. Set following attributes according your needs.
Attributes with defaults are optional and does not have to be set.

General attribute meaning
----
- A general command to FHEM looks like this
**\<command\> \<device\> \<reading\> \<value\>**

- e.g. 
**set MyLamp dim 75**

- widget attributes
**data-cmd data-device data-set data-set-on**

**receive data**
- data-get		: 		Reading name
- data-get-on   : 		Value for ON
- data-get-off  : 		Value for OFF

**send data**
- data-set		: 		Reading name
- data-set-on   : 		Value for ON
- data-set-off  : 		Value for OFF

####All widgets
- **data-type**      : widget type
- **data-device**    : FHEM device name (call FHEM's 'list' command to get all names)
- **class**          : CSS classes for look and formatting of the widget

####Switch widget
- **data-get**      : name of the reading to get from FHEM (default 'STATE')
- **data-set**      : name of the reading to set from FHEM (default '')
- **data-cmd**      : name of the command (**\<command\>** \<device\> \<reading\> \<value\>) (e.g. setstate, set, setreading, trigger) default: 'set'
- **data-doubleclick**: timeout to wait for x millisecondes click or touch. '0' disables the doubleclick feature. (default '0')
- **data-lock**  : name of the reading containing a boolean value for the lock (readonly) parameter (default <null>)
- **data-warn**  : name of the reading to be shown as a red warn overlay (default <null>)
- **data-warn-on**   : value(s) that show the warn badge (default 'true|on|[1-9]{1}[0-9]*')
- **data-warn-off**   : value(s) that hide the warn badge (default 'false|off|0')
- **data-warn-color**  : forecolor for warn badge (default '#aaa')
- **data-warn-background-color**  : background color for warn badge (default '#aa2200')
- **data-warn-icon**  : name of the font-awesome icon to be shown instead of the warn text. (default: '')
- **data-warn-fixed**  : number of digits after the decimal point (default: 0)
- **data-hide**   : name of the reading to hide/show the widget (default 'STATE')
- **data-hide-on**   : value for HIDE (default 'true|1|on')
- **data-hide-off**   : value for SHOW (default '!on')
- **data-hideparents**: jquery selector to hide element's parents too
- **data-reachable**  : name of the reading containing a integer value to indicate whether the device is reachable or not (default <null>)
- **data-timeout** .  : millisecondes until the switch falls back into the initial position (default: 0 [disabled])

dual state notation   
- **data-get-on**   : value for ON status to get or an array of states (default 'true|1|on|open|ON')
- **data-get-off**  : value for OFF status to get. (default 'false|0|off|closed|OFF')
- **data-set-on**   : value for ON status to set. (default: value of data-get-on)
- **data-set-off**  : value for OFF status to set. (default: value of data-get-off)
- **data-icon**     			: name of the font-awesome icon. (default: fa-lightbulb-o)
- **data-background-icon** 		: name of the font-awesome icon for background (default 'fa-circle')
- **data-on-background-color**          : color for ON state or DEVICE:READING for dynamic setting (default '#aa6900')
- **data-off-background-color**         : color for OFF state or DEVICE:READING for dynamic setting (default '#505050')
- **data-on-color** 			: color for ON state or DEVICE:READING for dynamic setting (default '#aa6900')
- **data-off-color**			: color for Off state or DEVICE:READING for dynamic setting (default '#505050')

multi state notation   
- **data-states**   			: array of states 
- **data-set-states**                   : array of states to set.
- **data-icons**    			: array of icons related to the data-states array 
- **data-background-icons**             : array of background icons related to the data-states array
- **data-colors**                       : array of colors related to the data-states array
- **data-background-colors**            : array of background colors related to the data-states array

- **class**     			: readonly, compressed, invert

data-get-on and data-get-off accept also RegEx values. e.g. data-get-on="[0-9]{1,3}|on" means set switch on if STATE is a numeric value or 'on'.
data-get-off="!on" means accept all but the data-get-on value (negation)
data-set-off="" suppress sending off

class 'invert' inverts foreground and background color

data-lock: the widget gets locked if the corresponding FHEM reading has the value 1, on or true.

data-[on|off]-color: the expected format for color value via reading is RGB (with ot without #)

See [examples](#switch) of Switch

####Symbol widget
- **data-get**      			: name of the reading to get from FHEM (default 'STATE')
- **data-warn**  : name of the reading to be shown as a red warn overlay (default <null>)
- **data-warn-on**   : value(s) that show the warn badge (default 'true|on|[1-9]{1}[0-9]*')
- **data-warn-off**   : value(s) that hide the warn badge (default 'false|off|0')
- **data-warn-color**  : forecolor for warn badge (default '#aaa')
- **data-warn-background-color**  : background color for warn badge (default '#aa2200')
- **data-hide**   : name of the reading to hide/show the widget (default 'STATE')
- **data-hide-on**   : value for HIDE (default 'true|1|on')
- **data-hide-off**   : value for SHOW (default '!on')
- **data-hideparents**: jquery selector to hide element's parents too
- **data-lock**  : name of the reading containing a boolean value for the lock (readonly) parameter (default <null>)
- **data-reachable**  : name of the reading containing a integer value to indicate whether the device is reachable or not (default <null>)


dual state notation
- **data-get-on**   			: value for ON status to get or an array of states (default 'true|1|on|open|ON')
- **data-get-off**  			: value for OFF status to get. (default 'false|0|off|closed|OFF')
- **data-icon**     			: name of the font-awesome icon.  (default 'ftui-window')
- **data-background-icon** 		: name of the font-awesome icon for background (default '')
- **data-on-background-color**          : fix color attribute for ON state or DEVICE:READING for dynamic setting (default '#aa6900')
- **data-off-background-color**         : fix color attribute for OFF state or DEVICE:READING for dynamic setting (default '#505050')
- **data-on-color** 			: fix color attribute for ON state or DEVICE:READING for dynamic setting (default '#aa6900')
- **data-off-color**			: fix color attribute for Off state or DEVICE:READING for dynamic setting (default '#505050')

multi state notation   
- **data-states**   			: array of states 
- **data-icons**    			: array of icons related to the data-states array 
- **data-background-icons**             : array of icons related to the data-states array
- **data-colors**                       : array of colors related to the data-states array
- **data-background-colors**            : array of colors related to the data-states array

- **class**     			: compressed

The CSS class 'compressed' forces the switch or symbol to a height and width of 1em instead of 2em. This saves space around the switch/symbol

data-get-on,data-get-off and data-states accept also RegEx values.
The value for one icon can also contain an additional animatation CSS name, e.g. "fa-exclamation-triangle fa-blink" for a blinking symbol

data-[on|off]-color: the expected format for color value via reading is RGB (with ot without #)

See [examples](#symbol) of Symbol

####Label widget
- **data-get**  : name of the reading containing label text
- **data-part** : RegEx or number (which word) for filtering shown text
- **data-fix**  : keeping a specified number of decimals. (default '-1' -> non-numeric)
- **data-factor**  : a numbers to be multiplied by the numeric reading value (default '' )
- **data-color**  : fix color attribute or DEVICE:READING for dynamic setting of label color (default '')
- **data-colors** : a array of color values. The used element for the label color defines the matching element of the limits array
- **data-classes**: a array of class names. The used element class defines the matching element of the limits array
- **data-limits** : a array of numeric or RegEx values for comparing with the current value
- **data-limits-get**  : name of the DEVICE:Reading to retrieve the value for comparing with the limits array (default: data-device:data-get)
- **data-limits-part**  : filter for the value. part number of the space separated value or an RegEx (default '-1' -> all)
- **data-unit** : add a unit after a numeric value.
- **data-substitution**: multiple functions to replace the original value (see descriptions below)
- **data-pre-text**: include this text before reading text
- **data-post-text**: include this text after reading text
- **data-hide**   : name of the reading to hide/show the widget (default 'STATE')
- **data-hide-on**   : value for HIDE (default 'true|1|on')
- **data-hide-off**   : value for SHOW (default '!on')
- **data-hideparents**: jquery selector to hide element's parents too
- **data-refresh**  : auto refresh interval in secondes (default '0', 0 means no auto refresh)
- **class**     : small, large, big, bigger, thin, red, green, blue, orange, darker, timestamp, w1x, w2x, w3x, circleborder
squareborder, bg-limit, icon, square, round, truncate

Use data-color OR data-colors + data-limits, not both.

With class="bg-limit" it changes the background color not the forecolor according data-limits

class="icon square" or class="icon round" forces the label to a fix width and height in icon style

Special layout can be achieved by overwriting of following classes in the fhem-tablet-ui-user.css:

.label-precomma
.label-comma
.label-aftercomma
.label-unit

e.g.:
.label-aftercomma{
    font-size:40%;
    left: 4px;
    top: -25px;
    position: relative;
}

Build-in classes are: aftercomma-top and unit-top

Functions for data-substitution:

1. Array of replacements
    e.g.: data-substitution='["on","Lampe ist an","off","Lampe ist aus"]'
2. RegEx-substitution to apply on the value. Standard regex notation (s/regex/subst/modifier) is expected
   e.g. data-substitution="s/no soundplayer active//g"
   
   data-substitution="s/(:00)$//g"                  -  20:15 instead of 20:15:00
   
   data-substitution="s/\b0*([1-9][0-9]*|0)\b/$1/g" -  4 instead of 0004 (removes leading zeros)
   
   
3. data-substitution="weekdayshort"
4. JS functions
data-substitution="toDate().ddmm()"             -  convert to day:month

data-substitution="toDate().hhmm()"             -  convert to hour:minutes

data-substitution="toDate().hhmmss()"           -  convert to hour:minutes:secondes

data-substitution="toDate().eeee()"             -  convert to name of the week day

data-substitution="toDate().ago()"              -  convert to time span (long format)

data-substitution="toDate().ago('hh:mm:ss')"    -  convert to time span (short format)

data-substitution="toString().toHoursFromSec()" -  convert to hours from secondes

data-substitution="addFactor(x)" - calculation (e.g. "addFactor(0.0001).toFixed(1)")

If you use class="timestamp" together with data-substitution="toDate().ago()" it is recommended to define data-refresh="xx" in secondes to refresh the value
from time to time, in case the reading timestamp refresh rate is low.

The use the label features only with a fixed label text, then use class="fixedlabel"

See [examples](#label) of Label

####Select widget
- **data-get**  : name of the reading that get the selected item of the list
- **data-set**  : name of the reading to set on FHEM (\<command\> \<device\> **\<reading\>** \<value\>) (default '')
- **data-list** : name of the reading to get a :-separated list from FHEM
- **data-items**: a array of fix items to show in the selection box and send to FHEM  (alternative if data-list is empty)
- **data-alias**: a array of fix names or a reading name which delivers such an array to show only in the selection box as an alias to the real items
- **data-cmd**  : name of the command to send to FHEM (**\<command\>** \<device\> \<reading\> \<value\>) (e.g. setstate, set, setreading, trigger) default: 'set'
- **data-quote** : characters to enclose the send value. (default '')
- **data-delimiter** : character which delimites list item. (default ':')
- **data-part** : RegEx or number (which word) for filtering the get reading
- **data-size** : number of visible options in the drop-down list. (default '1')

- **class**     : wider, w1x, w2x, w3x, large, big, notransmit

####Input widget
- **data-get**  : name of the reading that get the selected item of the list
- **data-set**  : name of the reading to set on FHEM (\<command\> \<device\> **\<reading\>** \<value\>) (default '')
- **data-cmd**  : name of the command to send to FHEM (**\<command\>** \<device\> \<reading\> \<value\>) (e.g. setstate, set, setreading, trigger) default: 'set'
- **data-value**: default value
- **class**     : wider, w1x, w2x, w3x, large, big, notransmit, autoclear, autoselect

data-device, data-get can be references (jQuery seletor) to select-widgets to change the source dynamically

####Push widget
- **data-set**    : name of the reading to set on FHEM (\<command\> \<device\> **\<reading\>** \<value\>) (default '')
- **data-set-on** : value (or an array of values) to send after the the button get released (or when pressed, if data-on-off is set) (default 'on')
- **data-set-off** : value to send after the the button get released. (default '') 
- **data-background-icon** : name of the font-awesome icon for background (default 'fa-circle')
- **data-cmd**  : name of the command (**\<command\>** \<device\> \<reading\> \<value\>) (e.g. setstate, set, setreading, trigger) default: 'set'
- **data-doubleclick**: timeout to wait for a second click or touch. '0' disables the doubleclick feature. (default '0')
- **data-countdown**: secondes for the countdown progress control (default: autodetect from 'on-for-timer' command)
- **data-icon**     			: name of the font-awesome icon.  (default 'ftui-window')
- **data-background-icon** 		: name of the font-awesome icon for background (default '')
- **data-off-background-color**         : fix color attribute for OFF state or DEVICE:READING for dynamic setting (default '#505050')
- **data-off-color**			: fix color attribute for Off state or DEVICE:READING for dynamic setting (default '#505050')
- **data-warn**  : name of the reading containing a integer value to be shown as a red warn overlay (default <null>)
- **data-hide**   : string to compare with current value. hide element when it's value equals data-hide. Or a DEVICE:READING to determine hiding from reading value  
- **data-hideparents**: jquery selector to hide element's parents too
- **data-lock**  : name of the reading containing a boolean value for the lock (readonly) parameter (default <null>)
- **data-reachable**  : name of the reading containing a integer value to indicate whether the device is reachable or not (default <null>)

'data-set-on' can also be an array of values to toggle between this values

####Knob widget
- **data-get**  : name of the reading containing the status value (default 'STATE')
- **data-set**  : name of the reading to set on FHEM (\<command\> \<device\> **\<reading\>** \<value\>) (default '')
- **data-cmd**  : name of the command (**\<command\>** \<device\> \<reading\> \<value\>) (e.g. setstate, set, setreading, trigger) default: 'set'
- **data-get-value** : RegEx to retrieve the value or part number of the space separated input to get the value (default '-1': all of the input)
- **data-set-value** : Format of the value to send to FHEM (default '$v': the value only)
- **data-min**  : minimal value to set (default 0)
- **data-max**  : maximal value to set (default 70)
- **data-initvalue**  : (default 10)
- **data-step**  :  (default 1);
- **data-angleoffset**  : (default -120);
- **data-anglearc**  : (default 240);
- **data-bgcolor**      : Color of background (default '#505050');
- **data-fgcolor**      : Color of ticks (default '#666');
- **data-nomcolor**   : Color of value  (default '#ffffff');
- **data-displayNominal**  : Show the value (default true);
- **data-font**  :  (default '"Helvetica Neue", "Helvetica", "Open Sans", "Arial", sans-serif');
- **data-font-weight**
- **data-unit** : add a unit after the center value.
- **data-lock**  : name of the reading containing the boolean value for the lock (readonly) parameter (default <null>)
- **class**		: mini, small, large, readonly

![](http://knowthelist.github.io/fhem-tablet-ui/knob.png)

####Thermostat widget
all parameters from knob widget plus following additional parameters
- **data-get**   : name of the reading containing the status value (default 'desired-temp')
- **data-temp**  : name of the reading for measured temperature of thermostates (default 'measured-temp')
- **data-set**   : name of the reading to set on FHEM (\<command\> \<device\> **\<reading\>** \<value\>) (default 'desired-temp')
- **data-valve** : name of the reading for valve position of thermostates
- **data-get-value** : RegEx to retrieve the value or part number of the space separated input to get the value (default '-1': all of the input)
- **data-mode**  : name of the reading for mode of thermostates
- **data-min**   : minimal value to set (default 10)
- **data-max**   : maximal value to set (default 30)
- **data-mincolor**   : Color of min temp (default '#4477ff');
- **data-maxcolor**   : Color of max temp (default '#ff0000');
- **data-actcolor**   : Color of current temp text (default '#999');
- **data-nomcolor**   : Color of value  (default '#ffffff');
- **data-step**  : step size for value adjustment e.g. 0.5 (default 1)
- **data-off**   : value to send to get the thermostat switch off (for this, dial the knob to then minimum value)
- **data-boost** : value to send to force boost mode (for this, dial the knob to then maximum value)
- **data-height**  : vertical size of the widget (default 100)
- **data-width**   : horizontal size of the widget (default 100)
- **data-touch-height**  : vertical size of the widget  during changes (default: the normal size - 100)
- **data-touch-width**   : horizontal size of the widget during changes (default the normal size - 100)
- **data-touch-position**   : position of the value during moving on touch devices (default 'left')


- **class**		 : mini, small, big, bigger, readonly

// special for MAX! WandThermostat
date-mode: if the value, retrieved from this reading equals 'auto' then such a command is created "set wz_WandThermostat desiredTemperature **auto** <value>".
Other values creates something like this "set wz_WandThermostat desiredTemperature <value>"

####Volume widget
all parameters from knob widget plus following additional parameters
- **data-get**  : name of the reading to get from FHEM (default 'STATE')
- **data-set**  : name of the reading to set on FHEM (\<command\> \<device\> **\<reading\>** \<value\>) (default '')
- **data-cmd**  : name of the command (**\<command\>** \<device\> \<reading\> \<value\>) (e.g. setstate, set, setreading, trigger) default: 'set'
- **data-get-value** : RegEx to retrieve the value or part number of the space separated input to get the value (default '-1': all of the input)
- **data-set-value** : Format of the value to send to FHEM (default '$v': the value only)
- **data-min**  : minimal value to set (default 0)
- **data-max**  : maximal value to set (default 70)
- **data-bgcolor**      : Color of background (default '#505050');
- **data-fgcolor**      : Color of ticks (default '#666');
- **data-nomcolor**   : Color of value  (default '#ffffff');
- **data-tickstep** : distance between ticks (default 4|20)
- **data-unit** : add a unit after the desired value.
- **class**		: mini, small, big, bigger, hue-tick, hue-front, hue-back, dim-tick ,dim-front, dim-back, readonly


class hue-tick		: draw ticks in color range
class hue-front		: draw handle in color range
class hue-back		: draw background in color range
class dim-tick 		: draw ticks in brightness range
class dim-front		: draw handle in brightness range
class dim-back		: draw background in brightness range

####Homestatus widget
all parameters from knob widget plus following additional parameters
- **data-get**      : name of the reading to get from FHEM (default 'STATE')

HISTORY
FHEM Tablet UI change history:

 2019-12-12
 - widget_slider: changed style classes
 - better websocket reconnecting

 2019-07-14
 - widget_tts: removed dependency of an external lib
 
 2019-06-05
 - fixed widget homestatus bug, that caused a number to appear in the middle

 2019-06-05
 - update of widget weather

 2019-05-28
 - enhanced dateFromString function
 
 2019-05-21
 - widget_calview: code optimization ; new attributes ; swiper fix
 
 2019-04-16
 - spinner: fixed a style bug
 - wdtimer: latest css version from eki 
 
 2019-04-02
 - weather: switched to version from somansch
 - wdtimer: latest version from forum 
 
 2019-03-17
 - removed WEB device query
 
 2019-02-18
 - slider: fixed a handle-position bug for bigger sizes
 - link: added support for embedded widgets

 2019-01-25
 - Fixed a bug that started initPage before getCSrf was finished
 
 2019-01-23
 - Fixed KW bug in clock widget
 
 2019-01-15
 - widget_link: added link parameter

 2018-12-12
 - fa-napster fixed
 
 2018-12-03
 - tts: update voice lib
 
 2018-12-01
 - image: refresh only if visible
 - pagebutton: fixed a bug which prevented forced reload
 
 2018-11-29
 - clock: corrected the serverDiff
 
 2018-11-21
 - checkbox: fixed the jack position issue
 
 2018-11-10
 - famultibutton: added warn-fixed
 
 2018-10-19
 - svgplot: fixed the src path
 
 2018-10-08
 - clock: added class check 'localonly'
 
 2018-10-07
 - clock: get time from FHEM, added data-offset
 
 2018-10-04
 - added minTime, maxTime to datetimepicker
 
 2018-10-02
 - ftui-circle-o
 - ftui-square-thin
 
 2018-09-28  (v2.7.6)
 - widget_playstream: improved
 - added addFactor() function
 
 2018-09-27 (v2.7.5)
 - added toHoursFromMin() function
 
 2018-09-21
 - widget_pagebutton: bugfix for just url buttons
 - widget_clock: corrects intervals < 100ms to 1000ms
 - widget_chart: corrections for fa-5 icons and more
 
 2018-09-19
 - jQuery 3.3.1
 - Font Awesome 5.3
 - JQuery-ui minimal
 
 2018-09-19
 - widget_weather: extented Proplanta-WeatherIcons map
 - widget_homestatus: optimized position for text and icons
 - widget_pagebutton: optimized return-timer and multistate support
 - widget_reload: new widget
 
 2018-09-05
 - widget_weather: added Proplanta-WeatherIcons map
 - widget_homestatus: fix for Fontawesome 5 icons

js/widget_volume.js
var widget_volume = {
  _volume: null,
  elements: null,
  rgbToHsl: function(rgb){
      var r=parseInt(rgb.substring(0,2),16);
      var g=parseInt(rgb.substring(2,4),16);
      var b=parseInt(rgb.substring(4,6),16);
        r /= 255, g /= 255, b /= 255;
        var max = Math.max(r, g, b), min = Math.min(r, g, b);
        var h, s, l = (max + min) / 2;

        if(max == min){
            h = s = 0; // achromatic
        }else{
            var d = max - min;
            s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
            switch(max){
                case r: h = (g - b) / d + (g < b ? 6 : 0); break;
                case g: h = (b - r) / d + 2; break;
                case b: h = (r - g) / d + 4; break;
            }
            h /= 6;
        }

        return [h, s, l];
  },
  hslToRgb: function(h, s, l){
      var r, g, b;
      var hex = function(x) {
        return ("0" + parseInt(x).toString(16)).slice(-2);
      }

      if(s == 0){
                 r = g = b = l; // achromatic
             }else{
                 function hue2rgb(p, q, t){
                     if(t < 0) t += 1;
                     if(t > 1) t -= 1;
                     if(t < 1/6) return p + (q - p) * 6 * t;
                     if(t < 1/2) return q;
                     if(t < 2/3) return p + (q - p) * (2/3 - t) * 6;
                     return p;
                 }

                 var q = l < 0.5 ? l * (1 + s) : l + s - l * s;
                 var p = 2 * l - q;
                 r = hue2rgb(p, q, h + 1/3);
                 g = hue2rgb(p, q, h);
                 b = hue2rgb(p, q, h - 1/3);
             }
        return [hex(Math.round(r * 255)), hex(Math.round(g * 255)), hex(Math.round(b * 255))].join('');
  },

  drawDial: function () {
  	var c = this.g, // context
	a = this.arc(this.cv), // Arc
	r = 1;

	c.lineWidth = this.lineWidth;
	c.lineCap = this.lineCap;
	if ((this.o.mode>>0) % 2 != 0) {
		this.o.bgColor='hsl('+ this.cv +',50%,50% )';
    } else if ((this.o.mode>>3) % 2 != 0) {
	    var cl=Math.floor(this.cv*(255/this.o.max));
	    this.o.bgColor = 'rgb('+cl+','+cl+','+cl+')';
	}
	if (this.o.bgColor !== "none") {
		c.beginPath();
		c.strokeStyle = this.o.bgColor;
		c.arc(this.xy, this.xy, this.radius, this.endAngle - 0.00001, this.startAngle + 0.00001, true);
		c.stroke();
	}
	
	var tick_w = (2 * Math.PI) / 360;
	var step =  (this.o.max - this.o.min) / this.angleArc;
	var acAngle = ((this.o.isValue - this.o.min) / step) + this.startAngle;
	var dist = this.o.tickdistance || 4;
	
	// draw ticks
	for (tick = this.startAngle; tick < this.endAngle + 0.00001; tick+=tick_w*dist) {
		i = step * (tick-this.startAngle)+this.o.min;
		
		c.beginPath();
		
		if ((this.o.mode>>1) % 2 != 0){
			// draw ticks in hue color
			c.strokeStyle = 'hsl('+ i +',50%,50% )';  
		} else if ((this.o.mode>>4) % 2 != 0){
		    var cl=Math.floor(i*(255/this.o.max));
		    c.strokeStyle = 'rgb('+cl+','+cl+','+cl+')';  
	    } else {
			// draw normal ticks
			c.strokeStyle = this.o.tkColor;//'#4477ff';
		}
		
		// thicker lines every 5 ticks
		if ( Math.round(i*10)/10 % 5 == 0 ){ 
			w = tick_w*2;
			w *= (c.strokeStyle != this.o.tkColor) ? 1.5 : 1; 
		}
		else {
			w = tick_w;
			w *= (c.strokeStyle != this.o.tkColor) ? 2 : 1;
		}
		// thicker lines every at current value
		//if (acAngle > tick-tick_w && acAngle < tick+tick_w)
			//w *= 1.9;	
			
		c.arc( this.xy, this.xy, this.radius, tick, tick+w , false);
		c.stroke();
	}

	// draw selection cursor
	c.beginPath();
	if ((this.o.mode>>2) % 2 != 0) {
		this.o.hdColor='hsl('+ this.cv +',50%,50% )';
	} else if ((this.o.mode>>5) % 2 != 0) {
	    var cl=Math.floor(this.cv*(255/this.o.max));
        this.o.hdColor = 'rgb('+cl+','+cl+','+cl+')';  
	}
		
	c.strokeStyle = this.o.hdColor;
	c.lineWidth = this.lineWidth * 2;
	c.arc(this.xy, this.xy, this.radius-this.lineWidth/2, a.s, a.e, a.d);
	c.stroke();

  return false;
  },
  init: function () {
  	_volume=this;
  	_volume.elements = $('div[data-type="volume"]');
 	_volume.elements.each(function(index) {
		var knob_elem =  jQuery('<input/>', {
			type: 'text',
			value: $(this).attr('data-initvalue')||'10',
			disabled : true,
		}).appendTo($(this));
		
		var device = $(this).data('device');
		$(this).data('get', $(this).data('get') || 'STATE');
		readings[$(this).data('get')] = true;
		
		var mode=0; //no hue colors
		var hdDefaultColor='#aa6900';
		if ($(this).hasClass('hue-back')){
			mode = mode | 1<<0;
			hdDefaultColor='#cccccc'; 
		}
		if ($(this).hasClass('hue-tick')){
			mode = mode | 1<<1; 
			hdDefaultColor='#bbbbbb';
		}
		if ( $(this).hasClass('hue-front')){
			mode = mode | 1<<2; 
		}
		
		if ($(this).hasClass('dim-back')){
			mode = mode | 1<<3;
		}
		if ($(this).hasClass('dim-tick')){
			mode = mode | 1<<4; 
		}
		if ( $(this).hasClass('dim-front')){
			mode = mode | 1<<5; 
		}
        if ( $(this).hasClass('rgb')){
            mode = mode | 1<<6;
        }

        $(this).data('mode',mode);

		var maxval = $(this).data('max') || 70;
		$(this).data('max', maxval);
		$(this).data('max360', (maxval>360)?360:maxval);
		
		$(this).data('height', 1*$(this).attr('data-height')||150);
        $(this).data('width', 1*$(this).attr('data-width')||150);
		if($(this).hasClass('small')) {
		    $(this).data('height', 100);
		    $(this).data('width', 100);
		}
		if($(this).hasClass('mini')) {
		    $(this).data('height', 52);
		    $(this).data('width', 52);
		}

		knob_elem.knob({
			'min': $(this).data('min') || 0,
			'max': $(this).data('max360'),
            'lastValue': 0,
            'variance': Math.floor($(this).data('max360') / 5),
			'origmax': $(this).data('max'),
            'height':$(this).data('height'),
            'width':$(this).data('width'),
			'angleOffset': $(this).attr('data-angleoffset')?$(this).attr('data-angleoffset')*1:-120,
			'angleArc': $(this).attr('data-anglearc')?$(this).attr('data-anglearc')*1:240,
			'bgColor': $(this).data('bgcolor') || 'transparent',
			'fgColor': $(this).data('fgcolor') || '#cccccc',
			'tkColor': $(this).data('tkcolor') || '#696969',
			'hdColor': $(this).data('hdcolor') || hdDefaultColor,
			'thickness': .25,
			'tickdistance': $(this).data('tickstep') || (((mode>>1) % 2 != 0)?4:20),
			'mode': mode,
			'cursor': 6,
			'cmd': $(this).data('cmd') || 'set',
			'set': $(this).data('set') || '',
			'draw' : _volume.drawDial,
			'readOnly' : $(this).hasClass('readonly'),
			'change' : function (v) { 
				  startInterval();
                    if (v > this.o.max - this.o.variance && this.o.lastValue < this.o.min + this.o.variance) {
                        knob_elem.val(this.o.min).change();
                        return false;
                    } else if (v < this.o.min + this.o.variance && this.o.lastValue > this.o.max - this.o.variance) {
                        knob_elem.val(this.o.max).change();
                        return false;
                    }
                    this.o.lastValue = v;
			},
			'release' : function (v) { 
				  if (ready){
                        if ((this.o.mode>>6) % 2 != 0){
                            //send hex rbg value
                            v=_volume.hslToRgb(v/this.o.max,1.0,0.5);
                        }
                        else{
                           //send decimal value
                           v=v*(this.o.origmax/this.o.max);
                           v=v.toFixed(0);
                        }
				  		if(typeof device!='undefined') {
                            var cmdl = [this.o.cmd,device,this.o.set,v].join(' ');
			  				setFhemStatus(cmdl);
			  				$.toast(cmdl);
				  		}
				  		this.$.data('curval', v);
				  }
			}	
		});
	 });
  },
  update: function (dev,par) {
	var deviceElements;
	if ( dev == '*' )
		deviceElements= _volume.elements;
	else
   		deviceElements= _volume.elements.filter('div[data-device="'+dev+'"]');

	deviceElements.each(function(index) {
		
		if ( $(this).data('get')==par || par =='*'){	
			var val = getDeviceValue( $(this), 'get' );
            var knob_elem = $(this).find('input');
			if (val){
				if ( knob_elem.val() != val ){
                    if ((parseInt($(this).data('mode'))>>6) % 2 != 0){
                        //is hex rgb
                        val=_volume.rgbToHsl(val)[0];
                        val=val*$(this).data('max360');
                    }
                    else{
                        //is deciaml value
                        val = (val * ($(this).data('max360')/$(this).data('max'))).toFixed(0);
                    }
					knob_elem.val( val ).trigger('change');
					DEBUG && console.log( 'volume dev:'+dev+' par:'+par+' change '+$(this).data('device')+':knob to ' +val );
				}	
			}
            knob_elem.css({visibility:'visible'});
		}
	});
   }
			 
};
